File: StartRemotePeers.java
import java.io.*;
import java.util.*;

/**
 * Reads PeerInfo.cfg and starts each peer process on its designated host via SSH.
 * For localhost testing, all peers are started locally.
 */
public class StartRemotePeers {

    public static void main(String[] args) {
        List<PeerInfo> peers = PeerInfo.loadAll("PeerInfo.cfg");
        String workingDir = System.getProperty("user.dir");

        for (PeerInfo peer : peers) {
            try {
                String host = peer.hostName;
                if (host.equals("localhost") || host.equals("127.0.0.1")) {
                    // Start locally
                    System.out.println("Starting peer " + peer.peerId + " locally...");
                    ProcessBuilder pb = new ProcessBuilder("java", "peerProcess", String.valueOf(peer.peerId));
                    pb.directory(new File(workingDir));
                    pb.redirectErrorStream(true);
                    Process p = pb.start();
                    // Read output in a separate thread
                    final int id = peer.peerId;
                    new Thread(() -> {
                        try (BufferedReader br = new BufferedReader(new InputStreamReader(p.getInputStream()))) {
                            String line;
                            while ((line = br.readLine()) != null) {
                                System.out.println("[Peer " + id + "] " + line);
                            }
                        } catch (IOException e) { /* ignore */ }
                    }).start();
                } else {
                    // Start via SSH
                    System.out.println("Starting peer " + peer.peerId + " on " + host + "...");
                    ProcessBuilder pb = new ProcessBuilder(
                        "ssh", host,
                        "cd " + workingDir + " && java peerProcess " + peer.peerId
                    );
                    pb.redirectErrorStream(true);
                    Process p = pb.start();
                    final int id = peer.peerId;
                    new Thread(() -> {
                        try (BufferedReader br = new BufferedReader(new InputStreamReader(p.getInputStream()))) {
                            String line;
                            while ((line = br.readLine()) != null) {
                                System.out.println("[Peer " + id + "] " + line);
                            }
                        } catch (IOException e) { /* ignore */ }
                    }).start();
                }
                Thread.sleep(500); // Small delay between starting peers
            } catch (Exception e) {
                System.err.println("Failed to start peer " + peer.peerId + ": " + e.getMessage());
            }
        }
    }
}

File: StartRemotePeers.py
import subprocess
import sys
import os
import time

from peer_info import PeerInfoEntry


def main():
    peers = PeerInfoEntry.load_all("PeerInfo.cfg")
    working_dir = os.getcwd()

    processes = []
    for peer in peers:
        host = peer.host_name
        if host in ("localhost", "127.0.0.1"):
            print(f"Starting peer {peer.peer_id} locally...")
            p = subprocess.Popen(
                [sys.executable, "peerProcess.py", str(peer.peer_id)],
                cwd=working_dir,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                text=True,
            )
            processes.append((peer.peer_id, p))
        else:
            print(f"Starting peer {peer.peer_id} on {host}...")
            p = subprocess.Popen(
                ["ssh", host, f"cd {working_dir} && python3 peerProcess.py {peer.peer_id}"],
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                text=True,
            )
            processes.append((peer.peer_id, p))
        time.sleep(0.5)

    # Print output from each process
    for peer_id, p in processes:
        p.wait()
        output = p.stdout.read()
        if output:
            for line in output.strip().splitlines():
                print(f"[Peer {peer_id}] {line}")


if __name__ == "__main__":
    main()
